staging/src/k8s.io/apiserver/pkg/server/config.go

```diff
@@ -1035,6 +1035,7 @@ func DefaultBuildHandlerChain(apiHandler http.Handler, c *Config) http.Handler {
 		handler = genericapifilters.WithTracing(handler, c.TracerProvider)
 	}
 	handler = genericapifilters.WithLatencyTrackers(handler)
+	handler = genericfilters.WithRoutine(handler, c.LongRunningFunc)
```

Comment:
Do you have any updated data if the gains from this PR changed after moving this here into the handler chain?

Reply:
Yes, 

```
# original
go_memstats_stack_inuse_bytes 2.0201472e+08

# after this PR
go_memstats_stack_sys_bytes 6.2816256e+07
```

Reply:
Hmm - you're saying this is improving usage by 70%?
That sounds a bit suspicious to me...

Reply:
Just to clarify, the 70% is from the whole PR, not from moving the handler chain. Moving the position of the handler chain only impacts 2 MiB or so, which might even be just statistical bias.

To elaborate on the 70% (ie. the 140MiB reduction), golang simply doubles the stack size if current stack size is exceeded, and due to the lengthy call depth in watch serving (also most of other resource serving calls as well), the [stack](https://github.com/kubernetes/kubernetes/issues/117777#issuecomment-1723055664) results in 32k in stack memory size for the primary goroutine for each watch connection. This RP reduces the stack memory to 4k or 2k (initial size), so the estimated memory reduction is (32k - 4k) * watch_connections (5k) = 140 MiB, which is perfectly matched what I observed from tests.

Reply:
I don't think so. Please see diff and full metrics (noting, the total usage diff is lower than claimed because Golang holds free memory and doesn't return back to system for various reasons)

Diff

```
diff /tmp/before /tmp/after
3c3
< go_memstats_alloc_bytes 5.53660392e+08
---
> go_memstats_alloc_bytes 5.32985304e+08
6c6
< go_memstats_alloc_bytes_total 2.3088914e+09
---
> go_memstats_alloc_bytes_total 2.383644024e+09
9c9
< go_memstats_buck_hash_sys_bytes 1.775686e+06
---
> go_memstats_buck_hash_sys_bytes 1.80191e+06
12c12
< go_memstats_frees_total 7.569195e+06
---
> go_memstats_frees_total 8.519271e+06
15c15
< go_memstats_gc_sys_bytes 2.3851392e+07
---
> go_memstats_gc_sys_bytes 2.4460152e+07
18c18
< go_memstats_heap_alloc_bytes 5.53660392e+08
---
> go_memstats_heap_alloc_bytes 5.32985304e+08
21c21
< go_memstats_heap_idle_bytes 1.27090688e+08
---
> go_memstats_heap_idle_bytes 1.6523264e+08
24c24
< go_memstats_heap_inuse_bytes 5.99801856e+08
---
> go_memstats_heap_inuse_bytes 6.5036288e+08
27c27
< go_memstats_heap_objects 5.08416e+06
---
> go_memstats_heap_objects 4.75143e+06
30c30
< go_memstats_heap_released_bytes 3.784704e+06
---
> go_memstats_heap_released_bytes 2.531328e+06
33c33
< go_memstats_heap_sys_bytes 7.26892544e+08
---
> go_memstats_heap_sys_bytes 8.1559552e+08
36c36
< go_memstats_last_gc_time_seconds 1.7029097843228054e+09
---
> go_memstats_last_gc_time_seconds 1.702909472212391e+09
42c42
< go_memstats_mallocs_total 1.2653355e+07
---
> go_memstats_mallocs_total 1.3270701e+07
51c51
< go_memstats_mspan_inuse_bytes 1.1601912e+07
---
> go_memstats_mspan_inuse_bytes 1.2004608e+07
54c54
< go_memstats_mspan_sys_bytes 1.1993856e+07
---
> go_memstats_mspan_sys_bytes 1.3216056e+07
57c57
< go_memstats_next_gc_bytes 1.049976896e+09
---
> go_memstats_next_gc_bytes 1.035159256e+09
60c60
< go_memstats_other_sys_bytes 3.952434e+06
---
> go_memstats_other_sys_bytes 3.930258e+06
63c63
< go_memstats_stack_inuse_bytes 1.95854336e+08
---
> go_memstats_stack_inuse_bytes 6.1014016e+07
66c66
< go_memstats_stack_sys_bytes 1.95854336e+08
---
> go_memstats_stack_sys_bytes 6.1014016e+07
69c69
< go_memstats_sys_bytes 9.64351448e+08
---
> go_memstats_sys_bytes 9.20049112e+08
```

Before

```
# HELP go_memstats_alloc_bytes Number of bytes allocated and still in use.
# TYPE go_memstats_alloc_bytes gauge
go_memstats_alloc_bytes 5.53660392e+08
# HELP go_memstats_alloc_bytes_total Total number of bytes allocated, even if freed.
# TYPE go_memstats_alloc_bytes_total counter
go_memstats_alloc_bytes_total 2.3088914e+09
# HELP go_memstats_buck_hash_sys_bytes Number of bytes used by the profiling bucket hash table.
# TYPE go_memstats_buck_hash_sys_bytes gauge
go_memstats_buck_hash_sys_bytes 1.775686e+06
# HELP go_memstats_frees_total Total number of frees.
# TYPE go_memstats_frees_total counter
go_memstats_frees_total 7.569195e+06
# HELP go_memstats_gc_sys_bytes Number of bytes used for garbage collection system metadata.
# TYPE go_memstats_gc_sys_bytes gauge
go_memstats_gc_sys_bytes 2.3851392e+07
# HELP go_memstats_heap_alloc_bytes Number of heap bytes allocated and still in use.
# TYPE go_memstats_heap_alloc_bytes gauge
go_memstats_heap_alloc_bytes 5.53660392e+08
# HELP go_memstats_heap_idle_bytes Number of heap bytes waiting to be used.
# TYPE go_memstats_heap_idle_bytes gauge
go_memstats_heap_idle_bytes 1.27090688e+08
# HELP go_memstats_heap_inuse_bytes Number of heap bytes that are in use.
# TYPE go_memstats_heap_inuse_bytes gauge
go_memstats_heap_inuse_bytes 5.99801856e+08
# HELP go_memstats_heap_objects Number of allocated objects.
# TYPE go_memstats_heap_objects gauge
go_memstats_heap_objects 5.08416e+06
# HELP go_memstats_heap_released_bytes Number of heap bytes released to OS.
# TYPE go_memstats_heap_released_bytes gauge
go_memstats_heap_released_bytes 3.784704e+06
# HELP go_memstats_heap_sys_bytes Number of heap bytes obtained from system.
# TYPE go_memstats_heap_sys_bytes gauge
go_memstats_heap_sys_bytes 7.26892544e+08
# HELP go_memstats_last_gc_time_seconds Number of seconds since 1970 of last garbage collection.
# TYPE go_memstats_last_gc_time_seconds gauge
go_memstats_last_gc_time_seconds 1.7029097843228054e+09
# HELP go_memstats_lookups_total Total number of pointer lookups.
# TYPE go_memstats_lookups_total counter
go_memstats_lookups_total 0
# HELP go_memstats_mallocs_total Total number of mallocs.
# TYPE go_memstats_mallocs_total counter
go_memstats_mallocs_total 1.2653355e+07
# HELP go_memstats_mcache_inuse_bytes Number of bytes in use by mcache structures.
# TYPE go_memstats_mcache_inuse_bytes gauge
go_memstats_mcache_inuse_bytes 19200
# HELP go_memstats_mcache_sys_bytes Number of bytes used for mcache structures obtained from system.
# TYPE go_memstats_mcache_sys_bytes gauge
go_memstats_mcache_sys_bytes 31200
# HELP go_memstats_mspan_inuse_bytes Number of bytes in use by mspan structures.
# TYPE go_memstats_mspan_inuse_bytes gauge
go_memstats_mspan_inuse_bytes 1.1601912e+07
# HELP go_memstats_mspan_sys_bytes Number of bytes used for mspan structures obtained from system.
# TYPE go_memstats_mspan_sys_bytes gauge
go_memstats_mspan_sys_bytes 1.1993856e+07
# HELP go_memstats_next_gc_bytes Number of heap bytes when next garbage collection will take place.
# TYPE go_memstats_next_gc_bytes gauge
go_memstats_next_gc_bytes 1.049976896e+09
# HELP go_memstats_other_sys_bytes Number of bytes used for other system allocations.
# TYPE go_memstats_other_sys_bytes gauge
go_memstats_other_sys_bytes 3.952434e+06
# HELP go_memstats_stack_inuse_bytes Number of bytes in use by the stack allocator.
# TYPE go_memstats_stack_inuse_bytes gauge
go_memstats_stack_inuse_bytes 1.95854336e+08
# HELP go_memstats_stack_sys_bytes Number of bytes obtained from system for stack allocator.
# TYPE go_memstats_stack_sys_bytes gauge
go_memstats_stack_sys_bytes 1.95854336e+08
# HELP go_memstats_sys_bytes Number of bytes obtained from system.
# TYPE go_memstats_sys_bytes gauge
go_memstats_sys_bytes 9.64351448e+08
```

After

```
# HELP go_memstats_alloc_bytes Number of bytes allocated and still in use.
# TYPE go_memstats_alloc_bytes gauge
go_memstats_alloc_bytes 5.32985304e+08
# HELP go_memstats_alloc_bytes_total Total number of bytes allocated, even if freed.
# TYPE go_memstats_alloc_bytes_total counter
go_memstats_alloc_bytes_total 2.383644024e+09
# HELP go_memstats_buck_hash_sys_bytes Number of bytes used by the profiling bucket hash table.
# TYPE go_memstats_buck_hash_sys_bytes gauge
go_memstats_buck_hash_sys_bytes 1.80191e+06
# HELP go_memstats_frees_total Total number of frees.
# TYPE go_memstats_frees_total counter
go_memstats_frees_total 8.519271e+06
# HELP go_memstats_gc_sys_bytes Number of bytes used for garbage collection system metadata.
# TYPE go_memstats_gc_sys_bytes gauge
go_memstats_gc_sys_bytes 2.4460152e+07
# HELP go_memstats_heap_alloc_bytes Number of heap bytes allocated and still in use.
# TYPE go_memstats_heap_alloc_bytes gauge
go_memstats_heap_alloc_bytes 5.32985304e+08
# HELP go_memstats_heap_idle_bytes Number of heap bytes waiting to be used.
# TYPE go_memstats_heap_idle_bytes gauge
go_memstats_heap_idle_bytes 1.6523264e+08
# HELP go_memstats_heap_inuse_bytes Number of heap bytes that are in use.
# TYPE go_memstats_heap_inuse_bytes gauge
go_memstats_heap_inuse_bytes 6.5036288e+08
# HELP go_memstats_heap_objects Number of allocated objects.
# TYPE go_memstats_heap_objects gauge
go_memstats_heap_objects 4.75143e+06
# HELP go_memstats_heap_released_bytes Number of heap bytes released to OS.
# TYPE go_memstats_heap_released_bytes gauge
go_memstats_heap_released_bytes 2.531328e+06
# HELP go_memstats_heap_sys_bytes Number of heap bytes obtained from system.
# TYPE go_memstats_heap_sys_bytes gauge
go_memstats_heap_sys_bytes 8.1559552e+08
# HELP go_memstats_last_gc_time_seconds Number of seconds since 1970 of last garbage collection.
# TYPE go_memstats_last_gc_time_seconds gauge
go_memstats_last_gc_time_seconds 1.702909472212391e+09
# HELP go_memstats_lookups_total Total number of pointer lookups.
# TYPE go_memstats_lookups_total counter
go_memstats_lookups_total 0
# HELP go_memstats_mallocs_total Total number of mallocs.
# TYPE go_memstats_mallocs_total counter
go_memstats_mallocs_total 1.3270701e+07
# HELP go_memstats_mcache_inuse_bytes Number of bytes in use by mcache structures.
# TYPE go_memstats_mcache_inuse_bytes gauge
go_memstats_mcache_inuse_bytes 19200
# HELP go_memstats_mcache_sys_bytes Number of bytes used for mcache structures obtained from system.
# TYPE go_memstats_mcache_sys_bytes gauge
go_memstats_mcache_sys_bytes 31200
# HELP go_memstats_mspan_inuse_bytes Number of bytes in use by mspan structures.
# TYPE go_memstats_mspan_inuse_bytes gauge
go_memstats_mspan_inuse_bytes 1.2004608e+07
# HELP go_memstats_mspan_sys_bytes Number of bytes used for mspan structures obtained from system.
# TYPE go_memstats_mspan_sys_bytes gauge
go_memstats_mspan_sys_bytes 1.3216056e+07
# HELP go_memstats_next_gc_bytes Number of heap bytes when next garbage collection will take place.
# TYPE go_memstats_next_gc_bytes gauge
go_memstats_next_gc_bytes 1.035159256e+09
# HELP go_memstats_other_sys_bytes Number of bytes used for other system allocations.
# TYPE go_memstats_other_sys_bytes gauge
go_memstats_other_sys_bytes 3.930258e+06
# HELP go_memstats_stack_inuse_bytes Number of bytes in use by the stack allocator.
# TYPE go_memstats_stack_inuse_bytes gauge
go_memstats_stack_inuse_bytes 6.1014016e+07
# HELP go_memstats_stack_sys_bytes Number of bytes obtained from system for stack allocator.
# TYPE go_memstats_stack_sys_bytes gauge
go_memstats_stack_sys_bytes 6.1014016e+07
# HELP go_memstats_sys_bytes Number of bytes obtained from system.
# TYPE go_memstats_sys_bytes gauge
go_memstats_sys_bytes 9.20049112e+08
```

Reply:
please add a comment explaining this,
also I commented here https://github.com/kubernetes/kubernetes/pull/120902#issuecomment-1861795535 but due the complexity introduced I may suggest have a feature flag to control the feature at least during one release

Reply:
Done